### DISCUSSION TITLE
Data typing for hive partitioning

### DISCUSSION THREAD

**Discussion by hawkfish** [duckdb/duckdb#7226] [Ideas]
It would be very convenient if we could add a column type argument when reading hive partitioned data.  For example, if I have data processing operation that produces a two level partitioning on date and run in the form `root/date=yyyy-mm-dd/run=n/schema.parquet` it would be simpler to be able to write

```sql
FROM read_parquet('root/*/*/schema.parquet', hive_partitioning=1, hive_types={'date':'DATE', 'ref':'SMALLINT'});
```

instead of 

```sql
SELECT * EXCLUDE(date, ref), date::DATE date, run::SMALLINT run
FROM read_parquet('root/*/*/schema.parquet', hive_partitioning=1);
```

The latter has to refer to each field three times (exclude, cast, as), which is very error prone.

**Comment by Mytherin** [comment]
This sounds like a great idea, and could also be combined with auto-detection of the types (for certain common types such as `DATE`, `TIMESTAMP`, `BIGINT`). @samansmink @lverdoes perhaps something to combine with the hive partitioning auto detection?

**Comment by Tishj** [comment]
I don't think we can do the proposed syntax:
```sql
hive_types={'date':'DATE', 'ref':'SMALLINT'});
```
Because the amount of children of the struct needs to be fixed, and in this case we would want them to be variable - which isn't possible

But instead we can do this:
```sql
hive_types={'names': ['date', 'ref'], 'types': ['DATE', 'SMALLINT']}
```

Actually nvm, if we set the named param to LogicalType::ANY it should be fine, we just need to add additional verification around it to make sure the value actually has the shape we expect
